src/blockchain/mempool_block_tree.cpp

```diff
@@ -1,49 +1,144 @@
 #include "veriblock/blockchain/blockchain_util.hpp"
 #include "veriblock/blockchain/mempool_block_tree.hpp"
+#include "veriblock/fmt.hpp"
 #include "veriblock/keystone_util.hpp"
 
 namespace altintegration {
 
-PopData VbkPayloadsRelations::toPopData() const {
-  PopData pop;
-  pop.context.push_back(*header);
-  for (const auto& vtb : vtbs) {
-    pop.vtbs.push_back(*vtb);
+bool MemPoolBlockTree::acceptVbkBlock(const std::shared_ptr<VbkBlock>& blk,
+                                      ValidationState& state) {
+  return temp_vbk_tree_.acceptBlock(blk, state);
+}
+
+bool MemPoolBlockTree::checkContextually(const ATV& atv,
+                                         ValidationState& state) {
+  // stateful validation
+  int32_t window = tree_->getParams().getEndorsementSettlementInterval();
+  auto duplicate = findBlockContainingEndorsement(
+      tree_->getBestChain(), tree_->getBestChain().tip(), atv.getId(), window);
+  if (duplicate) {
+    return state.Invalid(
+        "atv-duplicate",
+        fmt::sprintf("ATV=%s already added to active chain in block %s",
+                     atv.getId().toHex(),
+                     duplicate->toShortPrettyString()));
+  }
+
+  auto endorsed_hash =
+      tree_->getParams().getHash(atv.transaction.publicationData.header);
+  auto* endorsed_index = tree_->getBlockIndex(endorsed_hash);
+  if (endorsed_index != nullptr) {
+    auto* tip = tree_->getBestChain().tip();
+    assert(tip != nullptr && "block tree is not bootstrapped");
+
+    if (tip && (tip->getHeight() - endorsed_index->getHeight() + 1 > window)) {
```

Comment:
No need to check tip "if (tip)" here.
Add assert that endorsed_index height is not higher than tip height. Or avoid subtracting like: tip->getHeight() + 1 > window + endorsed_index->getHeight()

Reply:
fixed